lisa/tests/base.py

```diff
@@ -572,24 +573,72 @@ def get_rtapp_profile(cls, plat_info):
         pass
 
     @classmethod
-    def _run_rtapp(cls, target, res_dir, profile, ftrace_coll=None):
+    def get_cgroup_configuration(cls, plat_info):
+        """
+        :returns: a :class:`dict` representing the configuration of a cgroup
+          in a specific controller.
+
+        The returned dict must follow this example::
+
+         {
+             'controller': 'schedtune',
+             'name': 'lisa_test',
+             'values': {
+                 'boost': 10,
+                 'prefer_idle': 1
+             }
+         }
+
+        This will result in the creation of a cgroup named 'lisa_test' in the
+        'schedtune' controller, with a 'boost' value of 10 and 'prefer_idle'
+        value of 1. All rt-app tasks specified by get_rtapp_profile() will be
+        executed in the context of this cgroup.
+
+        This is a method you may optionally override to configure a cgroup for
+        the synthetic workload.
+        """
+        return None
+
+    @classmethod
+    def _run_rtapp(cls, target, res_dir, profile, cgroup=None, ftrace_coll=None):
         wload = RTA.by_profile(target, "rta_{}".format(cls.__name__.lower()),
                                profile, res_dir=res_dir)
 
         trace_path = os.path.join(res_dir, "trace.dat")
         ftrace_coll = ftrace_coll or FtraceCollector.from_conf(target, cls.ftrace_conf)
 
+        as_root = cgroup is not None
         with ftrace_coll, target.freeze_userspace():
-            wload.run()
+            wload.run(cgroup=cgroup, as_root=as_root)
         ftrace_coll.get_trace(trace_path)
         return trace_path
 
+    @classmethod
+    def _target_set_cgroup(cls, target, cgroup):
+        if not cgroup:
+            return None
+
+        # Get cgroup
+        ctrl_name = cgroup['controller']
+        ctrl = target.cgroups.controllers[ctrl_name]
+        cg_name = cgroup['name']
+        cg = ctrl.cgroup(cg_name)
+
+        # Set the values
+        for k, v in cgroup['values'].items():
+            cg.set(**{k: v})
+
+        return '/' + cg_name
+
     @classmethod
     def _from_target(cls, target, res_dir, ftrace_coll=None):
         plat_info = target.plat_info
         rtapp_profile = cls.get_rtapp_profile(plat_info)
-        cls._run_rtapp(target, res_dir, rtapp_profile, ftrace_coll)
+        cgroup_cfg = cls.get_cgroup_configuration(plat_info)
```

Comment:
In the current structure, it can be expected that:

* subclasses will override `_from_target()` completely, without calling `super()._from_target()`
* subclasses may "extend" `_run_rtapp()`, but will always call `super()._run_rtapp()` at some point.

As a consequence, if all this business can be done purely inside `RTATestBundle_run_rtapp()` and `get_cgroup_configuration()`, that should avoid creating any new parameter. 

Now that I'm thinking about it, having `rtapp_profile` as a `RTATestBundle.__init__` parameter is useless, since it can be fully derived from plat_info. We could remove the parameter and add a property if we want. We could follow the same pattern if the cgroups conf can be decided purely based on plat_info. That would allow "offline" code (test methods) to access that configuration, without actually recording it (since we record plat_info)

Reply:
Right, the problem is you can't assume the cgroup config can be directly re-constructed using `plat-info`, sometime it'll be hardcoded. That's already not the case for the schedtune tests for example.

But it is true that I have `cls.boost` and `cls.prefer_idle` already for `SchedTuneTest`, so the cgroup config dict is somewhat superfluous ... I'll drop it and see how things go if I merge everything in `_run_rtapp()` directly :wink:

Reply:
> Right, the problem is you can't assume the cgroup config can be directly re-constructed using plat-info, sometime it'll be hardcoded

Not sure to get it, as the current implementation seems to already take `plat_info` and just ignores it, which is perfectly fine. My point was that since it can be derived from either nothing (hardcoded) or something already saved (`plat_info`), we can just generate it on the fly in a property, so it becomes useless to pass it to `__init__` and we keep as "real" instance attributes the minimal state everything can be derived from.

The reason we pass `rtapp_profile` to `__init__()` is that it used to be built out of a Target, which is obviously not part of the bundle, so it had to be saved. Now that it is derived from `plat_info`, I propose to stop saving it as an attribute and just turn it into a property as well (saving the json in the artifacts for manual inspection is totally orthogonal to that and we will keep doing it regardless of what happens to that `rtapp_profile` attribute).

note: TestBundle attributes are the things that in the end are saved in the serialized objects for the tests. If something is only needed to create the bundle (i.e. in `from_target()` and its callees), and never accessed again in any instance method, there is no need to record it other than for manual inspection. That can be achieved by:

* exposing it as a getter method/property if it can be derived from existing attributes, 
* creating artifacts files
* If none of the above can be done, saving "unnecessary" attributes